packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+
+The table API to support this could look like the following:
+
+```TS
+public Table<TableRecord> extends FoundationElement {
+    ...
+
+    /*
+     * @internal
+     */
+    @observable
+    public readonly slottedColumns: TableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `TableCellData` interface
+
+A table cell represents a single column for a single row. The data that a cell has access to will be a subset of the data for the entire row. An instance of a table cell will be generic to describe the subset of data it contains, where the `TCellData` type is a superset of the type represented by [`TableRecord`](https://github.com/ni/nimble/blob/3e4b8d3dd59431d1671e381aa66052db57bc475c/packages/nimble-components/src/table/types.ts#L24):
+
+```TS
+interface TableCellState<TCellData extends TableRecord> {
+  data: TCellData;
+}
+```
+
+This interface could possibly be expanded in the future to communicate relevant table state to the cell template, such as whether or not the row is selected.
+
+### `TableColumn<>`
+
+This abstract class is what a column web component (i.e. a slotted column) must extend.
+
+```TS
+abstract class TableColumn<TCellData extends TableRecord = TableRecord> {
+    // This method will produce the expected TableCellState that the cellTemplate expects as
+    // its source
+    abstract generateCellState(cellData: TCellData): TableCellState<TCellData>;
+
+    // The template to use to render the cell content for the column
+    abstract cellTemplate: ViewTemplate<TableCellState<TCellData>>;
+
+    // The style to apply to the cellTemplate
+    cellStyles?: ElementStyles;
+
+    // The keys that should be present in TCellData.
+    // This array is parallel with the keys returned from `getDataKeys()`.
+    readonly cellDataKeyNames: readonly string[];
+
+    // The keys from the row data that correlate to the data that will be in TCellData.
+    // This array is parallel with the keys specified by `dataKeyNames`.
+    abstract getRowDataKeys(): string[];
+
+    // Function that allows the table column to validate the type that gets created
+    // for the cell data. This should validate that the types in TCellData are correct
+    // for each key defined by `dataKeyNames`.
+    // Return `true` if the data is valid. Return `false` if the data is not valid.
+    abstract validateCellData(cellData: TCellData): boolean;
```

Comment:
I think an exception might make sense here. It shouldn't be an expected code path right (definitely want to avoid on a hot path)? If data is poorly formatted according to the plugin which is implementing this method then we can't render.

If the contract is that this throws an exception then we can get nicer error reporting messages from the plugins. Thoguhts?

Reply:
I don't have a strong opinion either way. I'd be fine if we change this function to have a `void` return type and document it to throw an exception when the data is invalid.

Reply:
Updated to have this method indicate it should throw in invalid cases.